cmd/cdi/cmd/cdi-api.go

```diff
@@ -0,0 +1,351 @@
+	}
+}
+
+func cdiListDevices(verbose bool, format string) {
+	var (
+		registry = cdi.GetRegistry()
+		devices  = registry.DeviceDB().ListDevices()
+	)
+
+	if len(devices) == 0 {
+		fmt.Printf("No CDI devices found.\n")
+		return
+	}
+
+	fmt.Printf("CDI devices found:\n")
+	for idx, device := range devices {
+		cdiPrintDevice(idx, registry.DeviceDB().GetDevice(device), verbose, format, 2)
+	}
+}
+
+func cdiPrintDevice(idx int, dev *cdi.Device, verbose bool, format string, level int) {
+	if !verbose {
+		if idx >= 0 {
+			fmt.Printf("%s%d. %s\n", indent(level), idx, dev.GetQualifiedName())
+			return
+		}
+		fmt.Printf("%s%s\n", indent(level), dev.GetQualifiedName())
+		return
+	}
+
+	var (
+		spec = dev.GetSpec()
+	)
+
+	format = chooseFormat(format, spec.GetPath())
+
+	fmt.Printf("  %s (%s)\n", dev.GetQualifiedName(), spec.GetPath())
+	fmt.Printf("%s", marshalObject(level+2, dev.Device, format))
+	edits := spec.ContainerEdits
+	if len(edits.Env)+len(edits.DeviceNodes)+len(edits.Hooks)+len(edits.Mounts) > 0 {
+		fmt.Printf("%s global Spec containerEdits:\n", indent(level+2))
+		fmt.Printf("%s", marshalObject(level+4, spec.ContainerEdits, format))
+	}
+}
+
+func cdiShowSpecDirs() {
+	var (
+		registry  = cdi.GetRegistry()
+		specDirs  = registry.GetSpecDirectories()
+		cdiErrors = registry.GetErrors()
+	)
+	fmt.Printf("CDI Spec directories in use:\n")
+	for prio, dir := range specDirs {
+		fmt.Printf("  %s (priority %d)\n", dir, prio)
+		for path, specErrors := range cdiErrors {
+			if filepath.Dir(path) != dir {
+				continue
+			}
+			for _, err := range specErrors {
+				fmt.Printf("    - has error %v\n", err)
+			}
+		}
+	}
+}
+
+func cdiInjectDevices(format string, ociSpec *oci.Spec, patterns []string) error {
+	var (
+		registry = cdi.GetRegistry()
+		matches  = map[string]struct{}{}
+		devices  = []string{}
+	)
+
+	for _, device := range registry.DeviceDB().ListDevices() {
+		for _, glob := range patterns {
+			match, err := filepath.Match(glob, device)
+			if err != nil {
+				return errors.Wrapf(err, "failed to match pattern %q against %q",
+					glob, device)
+			}
+			if match {
+				matches[device] = struct{}{}
+			}
+		}
+	}
+	for device := range matches {
+		devices = append(devices, device)
+	}
+	sort.Strings(devices)
+
+	unresolved, err := registry.InjectDevices(ociSpec, devices...)
+
+	if len(unresolved) > 0 {
+		fmt.Printf("Unresolved CDI devices:\n")
+		for idx, device := range unresolved {
+			fmt.Printf("  %d. %s\n", idx, device)
+		}
+	}
+	if err != nil {
+		return errors.Wrapf(err, "OCI device injection failed")
+	}
+
+	fmt.Printf("Updated OCI Spec:\n")
+	fmt.Printf("%s", marshalObject(2, ociSpec, format))
+
+	return nil
+}
+
+func cdiResolveDevices(ociSpecFiles ...string) error {
+	var (
+		cache      *cdi.Cache
+		ociSpec    *oci.Spec
+		devices    []string
+		unresolved []string
+		err        error
+	)
+
+	if cache, err = cdi.NewCache(); err != nil {
+		return errors.Wrap(err, "failed to resolve devices")
```

Comment:
Nit: Is this error accurate? Which devices are being resolved at this point? Should this be `"failed to create CDI cache instance"` or something?

Reply:
Ouch... your're right, that error is bogus. Thanks for spotting !